File: src/modules/CartModule.js
/**
 * CartModule.js
 * Responsibility: Manage the items currently in the cart.
 */
export const CartModule = {
  cartItems: [],

  addItem(product, quantity = 1) {
    const existing = this.cartItems.find(item => item.id === product.id);
    if (existing) {
      existing.quantity += Number(quantity);
    } else {
      this.cartItems.push({ ...product, quantity: Number(quantity) });
    }
    return this.cartItems;
  },

  updateQuantity(productId, quantity) {
    const item = this.cartItems.find(i => i.id === productId);
    if (item) {
      item.quantity = Number(quantity);
    }
    return this.cartItems;
  },

  removeItem(productId) {
    this.cartItems = this.cartItems.filter(item => item.id !== productId);
    return this.cartItems;
  },

  getCartItems() {
    return this.cartItems;
  },

  calculateSubtotal() {
    return this.cartItems.reduce((total, item) => total + (item.price * item.quantity), 0);
  },

  clearCart() {
    this.cartItems = [];
  }
};

File: backend/CartModule.cpp
#include "CartModule.h"

void CartModule::addItem(const Product& product, int quantity) {
    for (auto& item : items) {
        if (item.product.id == product.id) {
            item.quantity += quantity;
            item.total = item.product.price * item.quantity;
            return;
        }
    }
    items.push_back({product, quantity, product.price * quantity});
}

void CartModule::updateQuantity(int productId, int quantity) {
    for (auto& item : items) {
        if (item.product.id == productId) {
            item.quantity = quantity;
            item.total = item.product.price * item.quantity;
            return;
        }
    }
}

void CartModule::removeItem(int productId) {
    for (auto it = items.begin(); it != items.end(); ++it) {
        if (it->product.id == productId) {
            items.erase(it);
            return;
        }
    }
}

std::vector<CartItem> CartModule::getItems() const {
    return items;
}

double CartModule::calculateSubtotal() const {
    double subtotal = 0;
    for (const auto& item : items) {
        subtotal += item.total;
    }
    return subtotal;
}

void CartModule::clear() {
    items.clear();
}
